In [1]:
import warnings
warnings.filterwarnings("ignore")

In [2]:
import tensorflow as tf

In [3]:
tf.__version__

'2.15.0'

In [4]:
if tf.config.list_physical_devices("GPU"):
    print("GPU")
else:
    print("CPU")

CPU


### constant

In [5]:
constant=tf.constant(42)
constant

<tf.Tensor: shape=(), dtype=int32, numpy=42>

In [6]:
constant.numpy()

42

In [7]:
type(constant)

tensorflow.python.framework.ops.EagerTensor

In [8]:
constant.dtype

tf.int32

In [9]:
constant1=tf.constant(42, dtype=tf.int64)
constant1

<tf.Tensor: shape=(), dtype=int64, numpy=42>

In [10]:
constant2= tf.constant([[2,3],[4,5]])
constant2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[2, 3],
       [4, 5]])>

#### Commonly used method to genereate constant tf.ones and tf.zeros just like np.ones and np.zeros

In [11]:
tf.ones(shape=(2,3))

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[1., 1., 1.],
       [1., 1., 1.]], dtype=float32)>

In [12]:
tf.zeros(shape=(3,2))

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[0., 0.],
       [0., 0.],
       [0., 0.]], dtype=float32)>

In [13]:
const1= tf.constant([[1,2],[2,3],[4,2]])
const2= tf.constant([[4,1],[1,2],[4,7]])
result=tf.add(const1,const2)
print(result)

tf.Tensor(
[[5 3]
 [3 5]
 [8 9]], shape=(3, 2), dtype=int32)


#### random constant

In [14]:
tf.random.normal(shape=(2,3),mean=0, stddev=1) #same as randn of numpy

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[ 0.4960703, -1.1527423, -1.3067404],
       [ 0.4395593, -0.3484977, -1.2489597]], dtype=float32)>

In [15]:
tf.random.uniform(shape=(2,3),minval=0, maxval=10, dtype=tf.int32)

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[2, 6, 8],
       [9, 1, 6]])>

### variable

A variable is a special tensor that is used to store the variable values and needs to be initialized with some values.

In [16]:
variable=tf.Variable([3,3])
variable

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([3, 3])>

In [17]:
type(variable)

tensorflow.python.ops.resource_variable_ops.ResourceVariable

In [18]:
var0=24
var1=tf.Variable(42)
var2=tf.Variable([[1,2],[2,3],[4.0,5.0]])
var0,var1,var2

(24,
 <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=42>,
 <tf.Variable 'Variable:0' shape=(3, 2) dtype=float32, numpy=
 array([[1., 2.],
        [2., 3.],
        [4., 5.]], dtype=float32)>)

#### The datatype can be explicitly specified

In [19]:
floatvar= tf.Variable(42, dtype=tf.float64)
floatvar

<tf.Variable 'Variable:0' shape=() dtype=float64, numpy=42.0>

#### To reassign a variable

In [20]:
var= tf.Variable(29.)
var

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=29.0>

In [21]:
var.assign(42.)

<tf.Variable 'UnreadVariable' shape=() dtype=float32, numpy=42.0>

In [22]:
initial_value= tf.random.normal(shape=(2,2))
a=tf.Variable(initial_value)
print(a)

<tf.Variable 'Variable:0' shape=(2, 2) dtype=float32, numpy=
array([[-1.6964247 ,  0.54196423],
       [-0.98091495,  0.29257286]], dtype=float32)>


#### Reshaping the tensor

In [23]:
tensor= tf.Variable([[[2,3],[2,5]],[[1,2],[3,4]]])
tensor

<tf.Variable 'Variable:0' shape=(2, 2, 2) dtype=int32, numpy=
array([[[2, 3],
        [2, 5]],

       [[1, 2],
        [3, 4]]])>

In [24]:
tensor1= tf.reshape(tensor,[2,-1])  
#the reshaping takes place like, if there is 8 data , it can be reshaped into 4*2, 2*4, 1*8, 8*1.
#That's why we use -1 in last , because it will automatically detect the shape.
tensor1

<tf.Tensor: shape=(2, 4), dtype=int32, numpy=
array([[2, 3, 2, 5],
       [1, 2, 3, 4]])>

In [25]:
tensor2= tf.reshape(tensor,[4,-1])  #It has automatically detect the 4*2.
tensor2

<tf.Tensor: shape=(4, 2), dtype=int32, numpy=
array([[2, 3],
       [2, 5],
       [1, 2],
       [3, 4]])>

#### Rank of Tensor

In [26]:
tensor

<tf.Variable 'Variable:0' shape=(2, 2, 2) dtype=int32, numpy=
array([[[2, 3],
        [2, 5]],

       [[1, 2],
        [3, 4]]])>

In [27]:
tf.rank(tensor)

<tf.Tensor: shape=(), dtype=int32, numpy=3>

#### specifying an element of a tensor   

In [28]:
tensor3= tensor[1,0,1] #slice 1, rows 0, column 1
tensor3

<tf.Tensor: shape=(), dtype=int32, numpy=2>

#### casting a tensor to a  Numpy Variable

In [29]:
tensor.numpy()

array([[[2, 3],
        [2, 5]],

       [[1, 2],
        [3, 4]]])

In [30]:
tensor[1,0,1].numpy()

2

#### Finding the size or length of a tensor

In [31]:
tensor_size= tf.size(input=tensor).numpy()
tensor_size

8

In [32]:
tensor.dtype

tf.int32

In [33]:
a= tf.random.normal(shape=(2,2))
b= tf.random.normal(shape=(2,2))
c= a+b
d=tf.square(c)
e=tf.exp(c)
print(a)
print(b)
print(c)
print(d)
print(e)

tf.Tensor(
[[ 1.5429881  -0.20382167]
 [-0.5830548  -0.963527  ]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[-0.33080027  0.54619575]
 [-1.121343   -0.28111276]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[ 1.2121878  0.3423741]
 [-1.7043978 -1.2446398]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[1.4693992  0.11722001]
 [2.9049718  1.5491282 ]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[3.3608294  1.408287  ]
 [0.18188189 0.28804466]], shape=(2, 2), dtype=float32)


In [34]:
tensor*tensor

<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[ 4,  9],
        [ 4, 25]],

       [[ 1,  4],
        [ 9, 16]]])>

In [35]:
tensor4= tensor*4 #broadcasting
tensor4

<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[ 8, 12],
        [ 8, 20]],

       [[ 4,  8],
        [12, 16]]])>

In [36]:
#transpose matrix multiplication 
tf.matmul(a, tf.transpose(b))

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[-0.6217474 , -1.672922  ],
       [-0.3333997 ,  0.92466414]], dtype=float32)>

In [37]:
#casting a tensor to another datatype 
tf.cast(tensor, dtype=tf.int64)

<tf.Tensor: shape=(2, 2, 2), dtype=int64, numpy=
array([[[2, 3],
        [2, 5]],

       [[1, 2],
        [3, 4]]], dtype=int64)>

In [38]:
tf.cast(tf.constant(4.4), dtype=tf.int64)

<tf.Tensor: shape=(), dtype=int64, numpy=4>

### Ragged Tensor

Ragged tensor is a type of tensor having one or more ragged dimension. Ragged dimensions are the dimension that have slices having various length. The simplest way of declearing a ragged tensor is declearning a ragged array. 

In [39]:
ragged= tf.ragged.constant([[[2,3,4,5],[],[4,5]],[[2,4],[4,5]]])
ragged

<tf.RaggedTensor [[[2, 3, 4, 5], [], [4, 5]], [[2, 4], [4, 5]]]>

In [40]:
#tf.constant([[[2,3,4,5],[],[4,5]],[[2,4],[4,5]]]) 
#this will show error because we cannot create a constant of different length.
#But using a ragged tensor we can create a constant of different length.

#### Squared differences

In [41]:
varx=[4,2,3,4] #returns (x-y)(x-y)
vary=2
varz=tf.math.squared_difference(varx,vary)
varz

<tf.Tensor: shape=(4,), dtype=int32, numpy=array([4, 0, 1, 4])>

#### Calculate a mean 

In [42]:
numbers= tf.constant([[1,2,3,4],[5,6,7,8]], dtype=tf.float64)
#add all the number and divide by the total no. of numbers.

In [43]:
tf.reduce_mean(input_tensor=numbers)

<tf.Tensor: shape=(), dtype=float64, numpy=4.5>

In [44]:
tf.reduce_mean(input_tensor=numbers, axis=0)  #calculate the mean across a column.

<tf.Tensor: shape=(4,), dtype=float64, numpy=array([3., 4., 5., 6.])>

In [45]:
tf.reduce_mean(input_tensor=numbers, axis=0, keepdims=True) #The reduced axis is retained into 1.

<tf.Tensor: shape=(1, 4), dtype=float64, numpy=array([[3., 4., 5., 6.]])>

In [46]:
tf.reduce_mean(input_tensor=numbers, axis=1) #calculate the mean across a column.

<tf.Tensor: shape=(2,), dtype=float64, numpy=array([2.5, 6.5])>

#### Random Variable Declearation

In [47]:
tf.random.normal(shape=(3,2),mean=10, stddev=2, seed=None, dtype=tf.float32, name=None)

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[11.973991 , 11.8844795],
       [12.399801 ,  6.366463 ],
       [ 7.605011 ,  8.8556   ]], dtype=float32)>

In [48]:
#Setting a seed
tf.random.set_seed(11)
tf.random.uniform(shape=(2,2),maxval=10,dtype=tf.int32)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[4, 6],
       [5, 2]])>

In [49]:
dice1=tf.random.normal([10,1], seed=11, dtype=tf.float32, name=None)
dice2=tf.random.normal([10,1], seed=11, dtype=tf.float32, name=None)
sum= dice1+dice2
tf.concat(values=[dice1,dice2,sum],axis=1)

<tf.Tensor: shape=(10, 3), dtype=float32, numpy=
array([[ 0.43841103,  2.4217978 ,  2.8602087 ],
       [ 1.391929  , -0.6691061 ,  0.7228229 ],
       [-0.91094244, -0.9223722 , -1.8333147 ],
       [-1.440929  , -0.02014393, -1.461073  ],
       [-1.7174875 ,  0.27540633, -1.4420811 ],
       [-2.3883555 ,  0.36795077, -2.0204048 ],
       [-0.8171607 ,  0.35199216, -0.46516857],
       [-0.27803615,  0.18207519, -0.09596096],
       [-0.96801126,  1.121964  ,  0.15395272],
       [-0.77267975,  0.5269573 , -0.24572247]], dtype=float32)>

#### Find the indexes of largest and smallest element

In [50]:
#1-d tensor
tensor1d= tf.constant([1,2,3,4,5,6,100,8,-1,-2,-3,-4])
print(tensor1d)

tf.Tensor([  1   2   3   4   5   6 100   8  -1  -2  -3  -4], shape=(12,), dtype=int32)


In [51]:
i= tf.argmax(input=tensor1d).numpy()
print("Index of min number :", i)
print("The number is:", tensor1d[i].numpy())

Index of min number : 6
The number is: 100


In [52]:
tf.argmin(input=tensor1d)

<tf.Tensor: shape=(), dtype=int64, numpy=11>

In [53]:
var1= tf.Variable([[1,2,3],[4,5,6]])
checkpoint=tf.train.Checkpoint(var=var1)
savepath=checkpoint.save("./vars")
var1.assign([[0,0,0],[0,0,0]])
var1

<tf.Variable 'Variable:0' shape=(2, 3) dtype=int32, numpy=
array([[0, 0, 0],
       [0, 0, 0]])>

In [54]:
checkpoint.restore(savepath)
var1

<tf.Variable 'Variable:0' shape=(2, 3) dtype=int32, numpy=
array([[1, 2, 3],
       [4, 5, 6]])>

#### Using tf.function

In [55]:
def f(x,y):
    return tf.reduce_mean(input_tensor=tf.multiply(x,y))
func= tf.function(f)
x=tf.constant(2)
y=tf.constant(4)

assert f(x,y).numpy() == func(x,y).numpy()

f(2,4)

<tf.Tensor: shape=(), dtype=int32, numpy=8>

#### Gradient Tape

In [56]:
a=tf.random.normal(shape=(2,2))
b=tf.random.normal(shape=(2,2))
with tf.GradientTape() as tape:
    tape.watch(a)
    c=tf.sqrt(tf.square(a)+tf.square(b))
    dc_da= tape.gradient(c,a)
    print(dc_da)

tf.Tensor(
[[ 0.4063939  0.6648737]
 [-0.9829512 -0.6326042]], shape=(2, 2), dtype=float32)


### Keras- A high level API for Tensorflow 2 

The Keras Sequential model
To build a Keras Sequential model, you add layers to it in the same order that you want the computations to be undertaken by the network.

After you have built your model, you compile it; this optimizes the computations that are to be undertaken, and is where you allocate the optimizer and the loss function you want your model to use.

The next stage is to fit the model to the data. This is commonly known as training the model, and is where all the computations take place. It is possible to present the data to the model either in batches, or all at once.

Next, you evaluate your model to establish its accuracy, loss, and other metrics. Finally, having trained your model, you can use it to make predictions on new data. So, the workflow is: build, compile, fit, evaluate, make predictions.

There are two ways to create a Sequential model. Let's take a look at each of them.

##### Using Sequential model
Firstly, you can pass a list of layer instances to the constructor, as in the following example.For now, we will just explain enough to allow you to understand what is happening here.

Acquire the data. MNIST is a dataset of hand-drawn numerals, each on a 28 x 28 pixel grid. Every individual data point is an unsigned 8-bit integer (uint8), as are the labels:

In [57]:
mnist_data = tf.keras.datasets.mnist
(train_x,train_y), (test_x, test_y) = mnist_data.load_data()

In [58]:
mnist_data

<module 'keras.api._v2.keras.datasets.mnist' from 'C:\\Users\\asus\\anaconda3\\lib\\site-packages\\keras\\api\\_v2\\keras\\datasets\\mnist\\__init__.py'>

#### Defining a variable

In [59]:
epochs=10
batch_size = 32

In [60]:
# normalize all the data points and cast the labels to int64
train_x, test_x = tf.cast(train_x/255.0, tf.float32), tf.cast(test_x/255.0, tf.float32)
train_y, test_y = tf.cast(train_y,tf.int64),tf.cast(test_y,tf.int64)

#### Building the Architecture

In [61]:
mnistmodel1 = tf.keras.models.Sequential([
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(512,activation=tf.nn.relu),
tf.keras.layers.Dropout(0.2),
tf.keras.layers.Dense(10,activation=tf.nn.softmax)
])

#### Compiling the model

In [62]:
optimiser = tf.keras.optimizers.Adam()
mnistmodel1.compile (optimizer= optimiser, loss='sparse_categorical_crossentropy', metrics = ['accuracy'])

#### Fitting the model

In [63]:
mnistmodel1.fit(train_x, train_y, batch_size=32, epochs=5)

Epoch 1/5


1875/1875 [==============================] - 12s 6ms/step - loss: 0.2167 - accuracy: 0.9355
Epoch 2/5
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0961 - accuracy: 0.9706
Epoch 3/5
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0690 - accuracy: 0.9784
Epoch 4/5
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0534 - accuracy: 0.9832
Epoch 5/5
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0425 - accuracy: 0.9862


#### Evaluate the mnistmodel1

In [64]:
mnistmodel1.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0694 - accuracy: 0.9795


[0.06941836327314377, 0.9794999957084656]


This represents a loss of 0.06 and an accuracy of 0.98 on the test data.

An accuracy of 0.98 means that out of 100 test data points, 98 were, on average, correctly identified by the model.